# Topic: Linear Regression Interpretation

## Definition (30-second explanation)
*   Linear regression models the linear relationship between a continuous dependent variable ($Y$) and one or more independent variables ($X$).
*   It estimates coefficients (betas) that describe the expected change in $Y$ for a one-unit change in $X$, assuming all other variables remain constant.
*   Ordinary Least Squares (OLS) finds these coefficients by minimizing the sum of squared residuals (the difference between predicted and actual values).

## Why Interviewers Ask This
*   To test your foundational statistical literacy and ensure you don't just blindly import `sklearn`.
*   To evaluate your ability to translate mathematical weights into actionable business insights (coefficient interpretation).
*   To check if you understand model validity guardrails, specifically the OLS assumptions, before deploying a model.

## Core Concepts
*   **Coefficients ($\beta_i$):** Represents the change in $Y$ for a 1-unit increase in $X_i$, holding all other features constant. 
*   **R-squared ($R^2$):** The proportion of variance in the target variable that is explained by the model (ranging from 0 to 1).
*   **Adjusted R-squared:** A modified $R^2$ that penalizes the addition of irrelevant features to prevent overfitting.
*   **LINE Assumptions:** The 4 pillars of OLS—Linearity, Independence (of errors), Normality (of residuals), and Homoscedasticity (constant variance of residuals).

## When to Use
*   When predicting a continuous outcome (e.g., house prices, salaries, monthly revenue).
*   As a baseline model before experimenting with more complex, black-box machine learning algorithms.
*   When strict interpretability is required by the business or regulatory bodies (e.g., explaining exactly why a credit limit was set).

## Advantages
*   **Highly Interpretable:** Stakeholders can easily understand the direct impact of each feature via its coefficient.
*   **Computationally Cheap:** Very fast to train and score, even on relatively large datasets.
*   **Statistical Inference:** Provides rich metrics (p-values, confidence intervals) to measure the statistical significance of relationships.

## Limitations
*   **Assumption Heavy:** Fails or provides misleading coefficients if the LINE assumptions are violated.
*   **Sensitive to Outliers:** Large outliers can heavily skew the best-fit line since OLS minimizes *squared* errors.
*   **Multicollinearity:** When features are highly correlated, coefficients become wildly unstable and lose their interpretability.

## Common Comparisons
*   **Linear vs. Logistic Regression:** Linear predicts a continuous target; Logistic predicts a binary/categorical probability.
*   **R-squared vs. Adjusted R-squared:** $R^2$ always increases (or stays the same) as you add features; Adjusted $R^2$ decreases if the new feature doesn't improve the model enough to justify the complexity.
*   **Sklearn vs. Statsmodels:** `sklearn` is optimized for machine learning and predictions; `statsmodels` is optimized for statistical inference and detailed summary reports.

## Common Interview Traps
*   **Interpreting the Intercept literally:** Explaining the intercept as a real-world baseline when $X=0$ is logically impossible (e.g., a house with 0 square feet).
*   **Confusing R-squared with Accuracy:** Stating that an $R^2$ of 0.7 means "70% of predictions are correct" (it actually means 70% of the *variance* is explained).
*   **Ignoring Assumptions:** Trusting a high $R^2$ without plotting residuals to check for heteroscedasticity or non-linearity.

## Python / SQL Syntax
```python
# 1. Prediction Focus (Sklearn)
from sklearn.linear_model import LinearRegression
model = LinearRegression().fit(X, y)
print(f"Coef: {model.coef_}, Intercept: {model.intercept_}")
print(f"R-squared: {model.score(X, y)}")

# 2. Inference Focus (Statsmodels)
import statsmodels.api as sm
X_sm = sm.add_constant(X) # Must manually add intercept
model_sm = sm.OLS(y, X_sm).fit()
print(model_sm.summary()) # Yields p-values, F-stat, Adj R-squared
```

## Important Formula
*   **Linear Equation:** $Y = \beta_0 + \beta_1 X_1 + ... + \beta_p X_p + \epsilon$
*   **R-squared:** $1 - \frac{SS_{res}}{SS_{tot}}$
*   **Adjusted R-squared:** $1 - \frac{(1-R^2)(n-1)}{n-p-1}$ *(where n=samples, p=features)*

## 45-Second Interview Answer
Linear regression models the linear relationship between a continuous target and its features. The most critical outputs are the coefficients, which tell us exactly how much the target changes for a one-unit change in a feature, holding everything else constant. However, for these coefficients to be statistically valid, the model must pass the four LINE assumptions: Linearity, Independence, Normality, and Homoscedasticity. In practice, I evaluate the model using Adjusted R-squared rather than standard R-squared, as it penalizes unnecessary complexity and helps prevent overfitting.

## Example Questions:

### Practice Question 1
**Question:** Your regression has R-squared=0.95 but Adjusted R-squared=0.60. What does this suggest?

**Ideal Interview Answer:**
This massive gap suggests the model is severely overfitted due to the inclusion of too many irrelevant features. Standard R-squared will mechanically increase every time a new feature is added, even if it's random noise. Adjusted R-squared penalizes the model for adding useless predictors. The drop to 0.60 indicates that a large portion of the features I included have no genuine predictive power for the target variable.

**Common Mistakes Candidates Make:**
*   Failing to explicitly mention the penalty for the number of predictors ($p$).
*   Assuming this indicates multicollinearity (it indicates over-parameterization/noise, not necessarily correlation between features).

**One Likely Interviewer Follow-up:**
"How would you go about fixing this model to bring the Adjusted R-squared closer to the R-squared?"

### Practice Question 2
**Question:** What happens to linear regression coefficients if you have multicollinearity?

**Ideal Interview Answer:**
When features are highly correlated, it becomes mathematically difficult for OLS to isolate the individual effect of each feature. As a result, the coefficients become highly unstable and standard errors inflate. A feature that is highly predictive might suddenly get a coefficient near zero, or even a sign flip (e.g., a positive predictor gets a negative coefficient). This destroys the model's interpretability. 

**Common Mistakes Candidates Make:**
*   Saying multicollinearity ruins the model's predictive accuracy (it usually doesn't harm overall prediction/$R^2$, it just ruins coefficient interpretability).
*   Not mentioning inflated standard errors or sign flipping.

**One Likely Interviewer Follow-up:**
"How do you detect multicollinearity, and how do you resolve it?" (Hint: VIF, Correlation Matrix / Ridge Regression, PCA, or dropping features).

### Practice Question 3
**Question:** What are the 4 assumptions of OLS linear regression? How do you check each?

**Ideal Interview Answer:**
The four assumptions form the acronym LINE:
1. **Linearity:** The relationship between X and Y is linear. Checked using a scatter plot of predicted values vs. actual values.
2. **Independence:** Errors are independent of one another. Checked using the Durbin-Watson statistic (especially important for time-series data).
3. **Normality:** The residuals (errors) are normally distributed. Checked using a Q-Q plot or a histogram of the residuals.
4. **Homoscedasticity:** The residuals have constant variance across all levels of X. Checked via a scatter plot of residuals vs. predicted values (looking for a random scatter, avoiding "funnel" shapes).

**Common Mistakes Candidates Make:**
*   Forgetting the acronym LINE.
*   Saying the *features* need to be normally distributed (it's the *residuals* that must be normal, not the input X).

**One Likely Interviewer Follow-up:**
"If you notice a distinct funnel shape in your residual plot, violating homoscedasticity, what transforms might you apply to fix it?"

### Practice Question 4
**Question:** How would you interpret a negative intercept in a house price model?

**Ideal Interview Answer:**
Mathematically, a negative intercept means the model predicts a negative house price when all features (like square footage, bedrooms, age) are exactly zero. However, in practical business terms, this interpretation is meaningless because a house with 0 square feet does not exist in reality. The intercept simply serves as a mathematical anchor to ensure the line of best fit minimizes errors across the actual, realistic data range. 

**Common Mistakes Candidates Make:**
*   Trying to invent a logical business reason for a negative price (e.g., "it means the land is toxic").
*   Stating that the model is broken or wrong because of the negative value.

**One Likely Interviewer Follow-up:**
"If the intercept is meaningless, should we just force the regression line to pass through the origin (intercept=0)?"

## Practice Questions:

### Q1: 
You are building a Linear Regression model to predict Customer Lifetime Value (CLV). The marketing team plans to use your coefficients to determine how much to spend acquiring different demographics. You notice a severe violation of **Normality** (a highly skewed Q-Q plot) and **Homoscedasticity** (a strong funnel shape in the residual plot). 

How do you explain the impact of these two specific violations to the non-technical marketing stakeholders who want to use your coefficients today? Secondly, what specific steps will you take to fix the model?

**Ideal Interview Answer:**
To the marketing team, I would explain that we cannot use the model today because our "confidence margins" are currently broken. The non-normality means our statistical significance tests are invalid—I might accidentally tell them a specific demographic is highly profitable when it's actually just random noise, leading to wasted marketing spend. The funnel shape (heteroscedasticity) means our model is highly inaccurate for our highest-value customers; the more a customer spends, the more wrong our model gets.

To fix this, CLV is typically heavily right-skewed. My first step would be to apply a log transformation to the target variable (`clv_target`) and re-train the model. This usually resolves both the normality and heteroscedasticity issues by compressing the large outliers. If heteroscedasticity persists, I would switch to calculating robust standard errors (like Huber-White) or use a Generalized Linear Model (GLM) like Poisson or Gamma regression instead of standard OLS.

**Common Mistakes Candidates Make:**
*   Explaining the problem mathematically (e.g., "the variance of epsilon is not constant") rather than in business terms (e.g., "we will misallocate budget").
*   Failing to mention a Log Transformation, which is the industry standard first step for skewed financial/revenue data.

**One Likely Interviewer Follow-up:**
"If you log-transform the target variable, how does that change the way you interpret the coefficients?" (Hint: It changes from an absolute unit increase to a percentage increase).

### Q2:
**Following up on the previous scenario, you decide to apply a log transformation to fix the assumption violations.**
Write the Python code using pandas and sklearn to:
- Apply a log transformation to the clv_target column (ensure you handle potential zero values safely).
- Define X (features) and your new y (target).
- Train a LinearRegression model and print the $R^2$ score.

In [ ]:
# Data:
import pandas as pd
import numpy as np

# Mock data to write your code against
data = {
    'age': [25, 34, 45, 23],
    'tenure_months': [12, 24, 36, 6],
    'total_purchases': [5, 15, 25, 2],
    'clv_target': [500, 1500, 3500, 200]
}
clv_df = pd.DataFrame(data)

In [9]:
import numpy as np
from sklearn.linear_model import LinearRegression

# 1. Apply log transformation safely using log1p (log(1 + x)) to handle zeros
clv_df['clv_target'] = clv_df['clv_target'].apply(np.log1p)

# 2. Define features (X) and target (y)
X = clv_df.drop('clv_target', axis=1)
y = clv_df['clv_target']

# 3. Train model and print R-squared
model = LinearRegression()
model.fit(X, y) # Note: Train/test split omitted for brevity
print(f'R2-Score: {model.score(X,y)}')

R2-Score: 1.0


**Common Mistakes Candidates Make:**
*   Using `np.log()` without checking for zero values, which throws an error or creates `-inf` values.
*   Forgetting to drop the target variable from the `X` dataframe, resulting in massive data leakage and an artificial $R^2$ of 1.0.

**One Likely Interviewer Follow-up:**
"When you predict a new value using this model, what mathematical operation must you apply to the output to get the actual dollar amount for CLV?" (Answer: `np.expm1()` to reverse the `log1p` transformation).

### Q3:
**You are analyzing the impact of different advertising channels on monthly revenue. The finance team isn't just looking for predictions; they want a detailed statistical breakdown to understand which channels are actually significant and what the confidence intervals are for their return on investment.**

Write the Python code to:
- Define your independent variables ($X$) and dependent variable ($y$).
- Prepare the features correctly for statsmodels (Hint: think about what sklearn does automatically that statsmodels requires you to do manually for linear regression).
- Fit an Ordinary Least Squares (OLS) model and print the full statistical summary.
- Follow-up question: Once that summary prints, which specific metric in the output table would you look at to confidently tell the finance team whether tv_spend is statistically significant, rather than just random noise?


In [15]:
# Data:
import pandas as pd
import statsmodels.api as sm

# Mock data
data = {
    'tv_spend': [230.1, 44.5, 17.2, 151.5, 180.8],
    'social_media_spend': [37.8, 39.3, 45.9, 41.3, 10.8],
    'monthly_revenue': [22.1, 10.4, 9.3, 18.5, 12.9]
}
marketing_df = pd.DataFrame(data)

In [16]:
marketing_df

,tv_spend,social_media_spend,monthly_revenue
0,230.1,37.8,22.1
1,44.5,39.3,10.4
2,17.2,45.9,9.3
3,151.5,41.3,18.5
4,180.8,10.8,12.9


In [17]:
import statsmodels.api as sm

# 1. Define features and target
X = marketing_df.drop('monthly_revenue', axis=1)
y = marketing_df['monthly_revenue']

# 2. Add constant (Statsmodels requires this manually to fit the intercept)
X_sm = sm.add_constant(X)

# 3. Fit OLS and print summary
model_ols = sm.OLS(y, X_sm).fit()
print(model_ols.summary())

                            OLS Regression Results                            
Dep. Variable:        monthly_revenue   R-squared:                       0.993
Model:                            OLS   Adj. R-squared:                  0.987
Method:                 Least Squares   F-statistic:                     150.0
Date:                Thu, 20 Aug 2026   Prob (F-statistic):            0.00662
Time:                        02:40:29   Log-Likelihood:                -2.4977
No. Observations:                   5   AIC:                             11.00
Df Residuals:                       2   BIC:                             9.824
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                 -1.8978      1

/home/shail/interview-prep/interview_env/lib/python3.12/site-packages/statsmodels/stats/stattools.py:74: ValueWarning: omni_normtest is not valid with less than 8 observations; 5 samples were given.
  warn("omni_normtest is not valid with less than 8 observations; %i "


To determine if `tv_spend` is statistically significant, I would look at its **p-value** (listed under `P>|t|` in the summary). The null hypothesis of OLS is that the feature's true coefficient is zero. If the p-value is less than 0.05, we reject the null hypothesis and confidently conclude the feature is statistically significant for predicting the target. Alternatively, I could check the 95% confidence interval `[0.025, 0.975]`; if the interval does not cross zero, the feature is significant.

**Common Mistakes Candidates Make:**
*   Forgetting `sm.add_constant(X)`, which forces the regression line through the origin (0,0) and drastically skews all coefficients and R-squared values.
*   Mixing up the input order. `statsmodels` expects `(y, X)`, whereas `sklearn` expects `(X, y)`. 

**One Likely Interviewer Follow-up:**
"In the `statsmodels` summary, you see both R-squared and Adjusted R-squared. If you added `radio_spend` to the model and R-squared went up, but Adjusted R-squared went down, what does that tell you about `radio_spend`?" (Answer: The feature does not improve the model enough to justify its inclusion; it is likely just noise).